### PR TITLE
XWALK-2697

### PR DESCRIPTION
**Pull request by ZhengXinCN**
Support for client-side SSL certificate for Android

Bug=XWALK-2697

### CODE REVIEW

**Comment by darktears** [comment]
Please squash all into one commit.

**Comment by wuhengzhi** [comment]
Test case is also needed for this feature. Thanks a lot for you patch. @ZhengXinCN

**Comment by ZhengXinCN** [comment]
Does anyone have any idea to do a unit test without a server requests Client Cert?

**Comment by ZhengXinCN** [comment]
@wuhengzhi  I added a simple test case that connects to this URL https://egov.privasphere.com/. I actually don't own this server, so not sure if this is a good idea. However, we need proper SSL certificates on this test server where we attempt to connect to and this one looks fine.

I am uncertain if the test was executed during "Testing patch series" and if it passed?

**Comment by wuhengzhi** [comment]
1. Please move the test case from XWalkCoreInternalTest to XWalkCoreTest, since the related function was overridden from XWalkResourceClient instead of XWalkResourceClientInternal.
2. Please enrich the commit log, e.g some code was ported from upstream; Add title for commit log: ```[Android] Support for client side SSL certificate.```
3. For test, please install XWalkCoreShell.apk, then execute ```./build/android/test_runner.py instrumentation --release --test-apk=XWalkCoreTest```

**Comment by wuhengzhi** [comment]
@ZhengXinCN From ```http://developer.android.com/reference/android/webkit/WebViewClient.html#onReceivedClientCertRequest(android.webkit.WebView, android.webkit.ClientCertRequest)```, we can see ```onReceivedClientCertRequest``` has only two paramters. 
I think we should supply the similar API to user. e.g. For XWalkCoreInternal, ```onReceivedClientCertRequest(XWalkViewInternal view,  ClientCertRequestHandlerInternal handler)```.
For XWalkCoreShell, ```onReceivedClientCertRequest(XWalkView view,  ClientCertRequestHandler handler)```.
This PR will also resolve XWALK-3667.

**Comment by ZhengXinCN** [comment]
@wuhengzhi OK, adapted as requested, can you please review once more?

**Comment by wuhengzhi** [comment]
I will review it today, thanks.

**Comment by wuhengzhi** [comment]
Please add ```BUG=XWALK-2697, XWALK-3667``` in comment.

**Comment by ZhengXinCN** [comment]
@wuhengzhi Thanks, I will fix them later on, and another question, do you have any code style formatter setting file for Eclipse of this?

**Comment by wuhengzhi** [comment]
Sorry, I have no format file.

**Comment by ZhengXinCN** [comment]
@wuhengzhi can you please review once more? when you have time. thanks

**Comment by wuhengzhi** [comment]
In the comment, ```wrer``` is ```were```?

**Comment by ahoward** [comment]
+1000

**Comment by wuhengzhi** [comment]
Thanks @ZhengXinCN, did you run test case by ```./build/android/test_runner.py instrumentation --release --test-apk=XWalkCoreTest```.
@axinging @lincsoon Please help to review, thanks.

**Comment by ZhengXinCN** [comment]
@wuhengzhi yes, I did run the test after I changed API, and my test case went pass. and I just test again seems no problem:

W    0.077s Main  org.xwalk.core.xwview.test.OnCreateWindowRequestedTest#testOnCreateWindowRequested has no annotations. Assuming "SmallTest".
W    0.212s Main  Creating 1 test runners.
W    0.213s bfe8  Creating shard 0 for device 4790f03ef70fbfe8.
W    2.115s bfe8  Unable to enable java asserts for 4790f03ef70fbfe8, non rooted device
W    4.880s Main  Running tests with 1 test runners.
W  194.828s bfe8  Will retry test org.xwalk.core.xwview.test.ShouldOverrideUrlLoadingTest#testCalledOnJavaScriptLocationDelayedReplaceRedirect, try #1.
W  469.244s bfe8  Test size not found in annotations for test 'org.xwalk.core.xwview.test.OnCreateWindowRequestedTest#testOnCreateWindowRequested', using 1 minute for timeout.
C  477.678s Main  ********************************************************************************
C  477.678s Main  Summary
C  477.678s Main  ********************************************************************************
C  477.679s Main  [==========] 124 tests ran.
C  477.679s Main  [  PASSED  ] 124 tests.
C  477.679s Main  ********************************************************************************

File: runtime/android/core_internal/src/org/xwalk/core/internal/XWalkResourceClientInternal.java [hunk 1]
```diff
@@ -270,4 +276,24 @@ public void onCancel(DialogInterface dialog) {
...
+     *
+     * This method is called on the UI thread. During the callback, the connection is suspended.
+     *
+     * The default behavior is to cancel, returning no client certificate.
+     *
+     * @param view The XWalkView that is initiating the callback
+     * @param handler An instance of a ClientCertRequestHandlerInternal
+     */
+    @XWalkAPI
+    public void onReceivedClientCertRequest(XWalkViewInternal view,
```
Comment (axinging) [inline runtime/android/core_internal/src/org/xwalk/core/internal/XWalkResourceClientInternal.java]: @lincsoon , this is API 5.0 or 6.0?

File: xwalk.gyp [hunk 2]
```diff
@@ -70,6 +70,7 @@
         'xwalk_resources',
         'extensions/extensions.gyp:xwalk_extensions',
         'sysapps/sysapps.gyp:sysapps',
+        '../third_party/boringssl/boringssl.gyp:boringssl',
```
Comment (axinging) [inline xwalk.gyp]: Why boring ssl imported?

**Comment by axinging** [comment]
I will give LGTM.  Thanks for Zhengxin's contribution. 
Please help to add one line in the new API  XWalkResourceClientInternal.onReceivedClientCertRequest's comments:

* @since 6.0

**Comment by lincsoon** [comment]
Why expose `ClientCertRequestHandlerInternal` to external? I couldn't see the need to extend `ClientCertRequestHandler` or any interface to handle a `ClientCertRequestHandler` object.

**Comment by lincsoon** [comment]
LGTM except the matter I mentioned above. Thanks for @ZhengXinCN 's contribution too!

**Comment by wuhengzhi** [comment]
```public void onReceivedClientCertRequest(XWalkView view,  ClientCertRequest handler)``` will be overriden bu user, so ```ClientCertRequestHandler``` need to expose to external. Please look at ```http://developer.android.com/reference/android/webkit/WebViewClient.html#onReceivedClientCertRequest(android.webkit.WebView, android.webkit.ClientCertRequest)```.

**Comment by lincsoon** [comment]
`ClientCertRequest` or `ClientCertRequestHandler`?

**Comment by wuhengzhi** [comment]
Oh, ```ClientCertRequestHandlerInternal``` is the internal implementation, need not to expose to external.

**Comment by ZhengXinCN** [comment]
seems I should expose android.webkit.ClientCertRequest.

**Comment by lincsoon** [comment]
I think it's okay to add the interface `org.xwalk.core.ClientCertRequest`. Just remove `ClientCertRequestHandler` from reflection generator if you don't want the developers to use it directly.

**Comment by ZhengXinCN** [comment]
@lincsoon 
if I removed "ClientCertRequestHandlerInternal" from generator, the generated class "XWalkResourceClientBridge.java" will failed to build (it wants [classname]Bridge).
 
User needs to override "onReceivedClientCertRequest" from "XWalkResourceClient" to handle client SSL request. Any idea to have a handler pass out without reflection generator?

**Comment by lincsoon** [comment]
Sorry, I didn't noticed that `ClientCertRequest` is an interface. Then you can keep `ClientCertRequestHandler` in the reflection generator.

But one more thing, you need put `ClientCertRequest` in the java docs so that the developer can know it. Please check the target `xwalk_core_library_documentation` in `xwalk_core_library_android.gypi` and add `ClientCertRequest` to the variable `api_files`. Then file the description of `ClientCertRequestInternal` just like `XWalkJavascriptResultInternal`, except the version is `since 6.0`.

**Comment by wuhengzhi** [comment]
```ClientCertRequestInternal``` could be declared as abstract class, please take https://github.com/wuhengzhi/crosswalk/commit/fc8e7f249e1884b0b21a7ee913dfe80fe0fe1079 for a reference.

**Comment by lincsoon** [comment]
Yes, you can change `ClientCertRequestInternal` to abstract class like `XWalkDownloadListener`, or keep current implementation like `XWalkJavascriptResult`. Because we don't have special provisions here, both styles are okay.

**Comment by ZhengXinCN** [comment]
@lincsoon I added it.
@wuhengzhi I prefer to use interface.

**Pull request merged**